core/jvm/src/main/scala/fs2/interop/flow/package.scala

```diff
@@ -0,0 +1,257 @@
+  *
+  * @example {{{
+  * scala> import cats.effect.{IO, Resource}
+  * scala> import fs2.Stream
+  * scala> import fs2.interop.flow._
+  * scala> import java.util.concurrent.Flow.Publisher
+  * scala>
+  * scala> val upstream: Stream[IO, Int] = Stream(1, 2, 3).covary[IO]
+  * scala> val publisher: Resource[IO, Publisher[Int]] = upstream.toPublisher
+  * scala> val downstream: Stream[IO, Int] = Stream.resource(publisher).flatMap { publisher =>
+  *      |   publisher.toStream[IO](bufferSize = 16)
+  *      | }
+  * scala>
+  * scala> import cats.effect.unsafe.implicits.global
+  * scala> downstream.compile.toVector.unsafeRunSync()
+  * res0: Vector[Int] = Vector(1, 2, 3)
+  * }}}
+  *
+  * @see [[java.util.concurrent.Flow]]
+  */
+package object flow {
+
+  /** Creates a [[Stream]] from a function exposing a [[Subscriber]].
+    *
+    * This function is useful when you actually need to provide a subscriber to a third-party.
+    * The subscription process is uncancelable.
+    *
+    * @example {{{
+    * scala> import cats.effect.IO
+    * scala> import fs2.Stream
+    * scala> import fs2.interop.flow.fromSubscriber
+    * scala> import java.util.concurrent.Flow.Subscriber
+    * scala>
+    * scala> // This internally calls somePublisher.subscribe(subscriber)
+    * scala> def thirdPartyLibrary(subscriber: Subscriber[Int]): Unit = ???
+    * scala>
+    * scala> // Interop with the third party library.
+    * scala> fromSubscriber[IO, Int](bufferSize = 16) { subscriber =>
+    *      |   IO.println("Subscribing!") >>
+    *      |   IO.delay(thirdPartyLibrary(subscriber)) >>
+    *      |   IO.println("Subscribed!")
+    *      | }
+    * res0: Stream[IO, Int] = Stream(..)
+    * }}}
+    *
+    * @note The publisher only receives a subscriber when the stream is run.
+    *
+    * @see [[fromPublisher]] for a simpler version that only requires a [[Publisher]].
+    *
+    * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+    *                   A high number can be useful if the publisher is triggering from IO,
+    *                   like requesting elements from a database.
+    *                   The publisher can use this `bufferSize` to query elements in batch.
+    *                   A high number will also lead to more elements in memory.
+    * @param onSubscribe The effectual function to run during the subscription process,
+    *                    receives a [[Subscriber]] that should be used to subscribe to a [[Publisher]].
+    *                    The `subscribe` operation must be called exactly once.
+    */
+  def fromSubscriber[F[_], A](
+      bufferSize: Int
+  )(
+      subscribe: Subscriber[A] => F[Unit]
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    Stream
+      .eval(StreamSubscriber[F, A](bufferSize))
+      .flatMap { subscriber =>
+        subscriber.stream(
+          F.uncancelable(_ => subscribe(subscriber))
```

Comment:
It should be up to the user to make this uncancelable. Since we are not chaining effects in the implementation here, this can be done completely (and more flexibly) in userland.
```suggestion
          subscribe(subscriber)
```